example/hooks/function_components.dart

```diff
@@ -0,0 +1,345 @@
+    } else {
+      print('count changed to ' + count.value.toString());
+      evenOdd.set('odd');
+    }
+    return () {
+      print('count is changing... do some cleanup if you need to');
+    };
+
+    /// This dependency prevents the effect from running every time [evenOdd.value] changes.
+  }, [count.value]);
+
+  return Fragment()(
+    (Dom.button()..onClick = (_) => count.set(1))('Reset'),
+    (Dom.button()
+      ..onClick = (_) => count.setWithUpdater((prev) => prev + 1))('+'),
+    Dom.br()(),
+    Dom.p()('${count.value} is ${evenOdd.value}'),
+  );
+},
+    FunctionComponentConfig(
+      displayName: 'UseStateTest',
+    ));
+
+Map initializeCount(int initialValue) {
+  return {'count': initialValue};
+}
+
+Map reducer(Map state, Map action) {
+  switch (action['type']) {
+    case 'increment':
+      return {...state, 'count': state['count'] + 1};
+    case 'decrement':
+      return {...state, 'count': state['count'] - 1};
+    case 'reset':
+      return initializeCount(action['payload']);
+    default:
+      return state;
+  }
+}
+
+mixin UseReducerTestProps on UiProps {
+  int initialCount;
+}
+
+UiFactory<UseReducerTestProps> UseReducerTest = uiFunction(
+  (props) {
+    final ReducerHook<Map, Map, int> state =
+        useReducerLazy(reducer, props.initialCount, initializeCount);
+
+    return Fragment()(
+      state.state['count'],
+      (Dom.button()
+        ..onClick = (_) => state.dispatch({'type': 'increment'}))('+'),
+      (Dom.button()
+        ..onClick = (_) => state.dispatch({'type': 'decrement'}))('-'),
+      (Dom.button()
+        ..onClick = (_) => state.dispatch({
+              'type': 'reset',
+              'payload': props.initialCount,
+            }))('reset'),
+    );
+  },
+  $UseReducerTestConfig, // ignore: undefined_identifier
+);
+
+UiFactory<UiProps> UseCallbackTest = uiFunction(
+  (props) {
+    final count = useState(0);
+    final delta = useState(1);
+
+    var increment = useCallback((_) {
+      count.setWithUpdater((prev) => prev + delta.value);
+    }, [delta.value]);
+
+    var incrementDelta = useCallback((_) {
+      delta.setWithUpdater((prev) => prev + 1);
+    }, []);
+
+    return Fragment()(
+      Dom.div()('Delta is ${delta.value}'),
+      Dom.div()('Count is ${count.value}'),
+      (Dom.button()..onClick = increment)('Increment count'),
+      (Dom.button()..onClick = incrementDelta)('Increment delta'),
+    );
+  },
+  FunctionComponentConfig(
+    displayName: 'UseCallbackTest',
+  ),
+);
+
+Context countContext = createContext(0);
+
+final UseContextTest = uiFunction<UiProps>(
+  (props) {
+    final count = useContext(countContext);
+
+    return Dom.div()(
+      Dom.div()(
+          'The count from context is $count'), // initially renders: 'The count from context is 0'
+    );
+  },
+  FunctionComponentConfig(displayName: 'UseContextTest'),
+);
+
+//final UseContextTest = uiFunction<UiProps>((props) {
+//  final context = useContext(TestNewContext);
```

Comment:
I'm not sure if this is part of this ticket, but this example doesn't work. I think it has something to do with the react-dart version of `useContext` returning `React.context` and here it returns the over_react version of `Context`. It also could have to do with me not knowing much about how `Context` works and converting the [react-dart example](https://github.com/cleandart/react-dart/blob/5.5.0-wip/example/test/function_component_test.dart#L168-L224) wrong.

Reply:
I was able to get to the bottom of this, I think! Looks like there was some missing typing on some of the calls into react-dart's context, unrelated to hooks.

The following changes got it working for me:
```patch
diff --git a/lib/src/util/context.dart b/lib/src/util/context.dart
index 40f7e7c5..2299e74f 100644
--- a/lib/src/util/context.dart
+++ b/lib/src/util/context.dart
@@ -49,7 +49,7 @@ class Context<TValue> {
   Context(this.Provider, this.Consumer, this.reactDartContext);
 
   /// The react.dart version of this context.
-  final react.Context reactDartContext;
+  final react.Context<TValue> reactDartContext;
 
   /// The react.js version of this context.
   ReactContext get jsThis => reactDartContext.jsThis;
@@ -237,7 +237,7 @@ class _DO_NOT_USE_OR_YOU_WILL_BE_FIRED {
 ///
 /// Learn more: <https://reactjs.org/docs/context.html#reactcreatecontext>
 Context<TValue> createContext<TValue>([TValue defaultValue, int Function(TValue, TValue) calculateChangedBits]) {
-  react.Context reactDartContext = react.createContext(defaultValue, calculateChangedBits != null ? (dynamic arg1, dynamic arg2) => calculateChangedBits(arg1, arg2) : null);
+  final reactDartContext = react.createContext<TValue>(defaultValue, calculateChangedBits != null ? (dynamic arg1, dynamic arg2) => calculateChangedBits(arg1, arg2) : null);
   // ignore: prefer_function_declarations_over_variables, avoid_types_on_closure_parameters
   UiFactory<ProviderProps> Provider = ([Map map]) => (ProviderProps<TValue>(map)..componentFactory = reactDartContext.Provider);
   // ignore: prefer_function_declarations_over_variables, avoid_types_on_closure_parameters
```

Are you cool just including those fixes in this PR? I think they're pretty low-risk, and are definitely considered bug fixes

Reply:
Yay!!! Thank you! I'll add that fix